From bc566d42e6b29534c62d3c814b6db5c2be2cc14a Mon Sep 17 00:00:00 2001
From: Kumar Gala <galak@kernel.crashing.org>
Date: Fri, 5 Mar 2010 13:48:19 -0600
Subject: [PATCH 111/233] of_mdio: Add of_phy_attach function

Extracted from vendor drop QorIQ-DPAA-SDK-V1-20110609-systembuilder.iso
0140-of_mdio-Add-of_phy_attach-function.patch

of_phy_connect is useful for most systems, but some drivers will want
finer-grained control over their PHYs, and won't want to use the
PHY Lib state machine or interrupt handlers.

Signed-off-by: Andy Fleming <afleming@freescale.com>
---
 drivers/of/of_mdio.c    |   14 ++++++++++++++
 include/linux/of_mdio.h |    3 +++
 2 files changed, 17 insertions(+), 0 deletions(-)

diff --git a/drivers/of/of_mdio.c b/drivers/of/of_mdio.c
index b474833..c169e99 100644
--- a/drivers/of/of_mdio.c
+++ b/drivers/of/of_mdio.c
@@ -180,3 +180,17 @@ struct phy_device *of_phy_connect_fixed_link(struct net_device *dev,
 	return IS_ERR(phy) ? NULL : phy;
 }
 EXPORT_SYMBOL(of_phy_connect_fixed_link);
+
+/* XXX add comment */
+struct phy_device *of_phy_attach(struct net_device *dev,
+				struct device_node *phy_np, u32 flags,
+				phy_interface_t iface)
+{
+	struct phy_device *phy = of_phy_find_device(phy_np);
+
+	if (!phy)
+		return NULL;
+
+	return phy_attach_direct(dev, phy, flags, iface) ? NULL : phy;
+}
+EXPORT_SYMBOL(of_phy_attach);
diff --git a/include/linux/of_mdio.h b/include/linux/of_mdio.h
index 53b94e0..514fef4 100644
--- a/include/linux/of_mdio.h
+++ b/include/linux/of_mdio.h
@@ -21,5 +21,8 @@ extern struct phy_device *of_phy_connect(struct net_device *dev,
 extern struct phy_device *of_phy_connect_fixed_link(struct net_device *dev,
 					 void (*hndlr)(struct net_device *),
 					 phy_interface_t iface);
+extern struct phy_device *of_phy_attach(struct net_device *dev,
+				struct device_node *phy_np, u32 flags,
+				phy_interface_t iface);
 
 #endif /* __LINUX_OF_MDIO_H */
-- 
1.7.0.4

